platforms/software/security/src/main/java/org/gradle/security/internal/pgp/BaseInMemoryPgpSignatoryProvider.java

```diff
@@ -44,24 +45,23 @@ public class BaseInMemoryPgpSignatoryProvider implements BaseSignatoryProvider<P
 
     private final PgpSignatoryFactory factory = new PgpSignatoryFactory();
     private final Map<String, PgpSignatory> signatories = new LinkedHashMap<>();
-    private final Lazy<PgpSignatory> defaultSignatory;
+    private final Provider<PgpSignatory> defaultSignatory;
 
-    public BaseInMemoryPgpSignatoryProvider(String defaultSecretKey, String defaultPassword) {
-        this(null, defaultSecretKey, defaultPassword);
-    }
+    public BaseInMemoryPgpSignatoryProvider(Provider<String> defaultKeyId, Provider<String> defaultSecretKey, Provider<String> defaultPassword) {
+        this.defaultSignatory = defaultKeyId.zip(defaultSecretKey.zip(defaultPassword, Pair::of), (keyId, secretKeyPasswordPair) -> {
```

Comment:
I never used a `ValueSource`, wouldn't be this a good candidate to implement the caching?

Reply:
> I never used a `ValueSource`, wouldn't be this a good candidate to implement the caching?

It caches the obtained value internally, but will not help with changing inputs. I suspect it can also impose performance penalty to the configuration cache fingerprinting, though I don't recall exactly if the signatory is being obtained at configuration time at all, probably not.